### ISSUE TITLE
hook first layer problem

### ISSUE THREAD

**Issue by shijianjian** [utkuozbulak/pytorch-cnn-visualizations#67]
Hi,

Thanks for this splendid repo. I am trying to generate feature CAMs from backprop based methods for EffientNet model, while I see the code below.
```python
    def hook_layers(self):
        def hook_function(module, grad_in, grad_out):
            self.gradients = grad_in[0]

        # Register hook to the first layer
        first_layer = list(self.model._modules.items())[0][1]
        first_layer.register_backward_hook(hook_function)
```
I acctually got ```self.gradeients = None``` while ```grad_in = (None, torch.tensor(32, 3, 3, 3), None)```. According to [here](https://discuss.pytorch.org/t/what-do-the-grad-in-and-grad-out-of-nn-conv2d-consist-of/41867/3). The ```grad_in``` will recieve (input, weight, bias), while the input will be None for the first layer.

Hence, from my understanding, the grad_in[0] for the first layer does not have the gradients for the input. In this case, how can I have the input gradients for the task?

Thanks in advance.

**Comment by stefanopini** [comment]
Hi @shijianjian ,

I've just encountered the same issue (using pytorch 1.3).
After looking at this [pytorch issue (#598)](https://github.com/pytorch/pytorch/issues/598) and pytorch docs, I changed from using `register_backward_hook` on the first network module to using `register_hook` on the input tensor.

You can find the modified code in pull request #68.
Hope @utkuozbulak can check if it's correct and include it in this repository.

**Comment by shijianjian** [comment]
Hi @utkuozbulak,

Thanks for the solution. My current workaround is to update the input image tensor to ```input.requires_grad=True```.  It works with the current ```register_backward_hook ``` implementation.

Although I haven't tested on your solution, I think ```register_hook``` may be better than the ```register_backward_hook``` as it is more concise and straightforward.

**Comment by utkuozbulak** [comment]
Hello @stefanopini, thanks for the PR. I will have a look at it in the upcoming days.

**Comment by utkuozbulak** [comment]
I'm glad @stefanopini found solution for your problem. See the comment at #68 for my reply for the pull request.
